maro/simulator/scenarios/vm_scheduling/business_engine.py

```diff
@@ -0,0 +1,482 @@
+            self._config = DottableDict(safe_load(fp))
+
+        self._delay_duration: int = self._config.DELAY_DURATION
+        self._buffer_time_budget: int = self._config.BUFFER_TIME_BUDGET
+        self._pm_amount: int = self._config.PM["P1"]["AMOUNT"]
+
+    def _init_data(self):
+        vm_table_data_path = self._config.VM_TABLE
+        if vm_table_data_path.startswith("~"):
+            vm_table_data_path = os.path.expanduser(vm_table_data_path)
+
+        cpu_readings_data_path = self._config.CPU_READINGS
+        if cpu_readings_data_path.startswith("~"):
+            cpu_readings_data_path = os.path.expanduser(cpu_readings_data_path)
+
+        if (not os.path.exists(vm_table_data_path)) or (not os.path.exists(cpu_readings_data_path)):
+            self._download_processed_data()
+
+    def _init_pms(self):
+        """Initialize the physical machines based on the config setting. The PM id starts from 0."""
+        self._pm_cpu_cores_capacity: int = self._config.PM["P1"]["CPU"]
+        self._pm_memory_capacity: int = self._config.PM["P1"]["MEMORY"]
+
+        # TODO: Improve the scalability. Like the use of multiple PM sets.
+        self._machines = self._frame.pms
+        for pm_id in range(self._pm_amount):
+            pm = self._machines[pm_id]
+            pm.set_init_state(
+                id=pm_id,
+                cpu_cores_capacity=self._pm_cpu_cores_capacity,
+                memory_capacity=self._pm_memory_capacity
+            )
+
+    def reset(self):
+        """Reset internal states for episode."""
+        self._total_energy_consumption: float = 0.0
+        self._successful_placement: int = 0
+        self._successful_completion: int = 0
+        self._failed_placement: int = 0
+        self._total_latency: Latency = Latency()
+        self._total_oversubscriptions: int = 0
+
+        for pm in self._machines:
+            pm.reset()
+
+        self._live_vms: Dict[int, VirtualMachine] = {}
+        self._pending_vm_request_payload: Dict[int, VmRequestPayload] = {}
+
+        self._frame.reset()
+
+        self._snapshots.reset()
+
+        self._vm_reader.reset()
+        self._vm_item_picker = self._vm_reader.items_tick_picker(self._start_tick, self._max_tick, time_unit="s")
+
+        self._cpu_reader.reset()
+        self._cpu_utilization_dict: Dict[int, float] = {}
+
+    def _init_frame(self):
+        self._frame = build_frame(self._pm_amount, self.calc_max_snapshots())
+        self._snapshots = self._frame.snapshots
+
+    def step(self, tick: int):
+        """Push business to next step.
+
+        Args:
+            tick (int): Current tick to process.
+        """
+        self._tick = tick
+        # Process finished VMs.
+        self._process_finished_vm()
+        # Update all live VMs CPU utilization.
+        self._update_vm_workload()
+        # Update all PM CPU utilization.
+        self._update_pm_workload()
+
+        self._cpu_utilization_dict = self._cpu_reader.items(tick=tick)
```

Comment:
As mentioned in a comment added previously, no need to main the ``_cpu_utilization_dict`` as a class variable. You can call ``cpu_reader.items()`` before ``process_finished_vm``, and deliver it to ``update_pm_workload`` as a function parameter.